src/dataflow/src/render/join.rs

```diff
@@ -46,46 +50,90 @@ where
             implementation: expr::JoinImplementation::Differential((start, start_arr), order),
         } = relation_expr
         {
-            let column_types = relation_expr.typ().column_types;
-            let arity = column_types.len();
+            let input_mapper = expr::JoinInputMapper::new(inputs);
+            let output_arity = input_mapper.total_columns();
+
+            // Determine dummy columns for un-demanded outputs, and a projection.
+            let (dummies, demand_projection) = if let Some(demand) = demand {
+                let mut dummies = Vec::new();
+                let mut demand_projection = Vec::new();
+                for (column, typ) in relation_expr.typ().column_types.into_iter().enumerate() {
+                    if demand.contains(&column) {
+                        demand_projection.push(column);
+                    } else {
```

Comment:
I just want to make sure I'm understanding this right

lets say theres two input columns `[0, 1]` and 1 is demanded and zero is not. then at the end of this for loop

demand_projection should look like `[2, 1]`

and dummies should look like `[ScalarExpr::Dummy]`

and then when you call `project` with `demand_project` you _have_ to have called the `map` previously, because what its saying is "keep the 3rd and second projections" in that order. is that right?

Reply:
Right. And, btw, this is the thing that `delta_join.rs` does not do (pre-load a MFP that projects away things that aren't needed). It could, but it didn't previously so I didn't add it.

Reply:
if thats alright then I wonder if maybe its also worth it to force a `remove_undemanded` call after line 77? it seems like we could replace some input (in this case the first input) with a dummy, but still try to compute something from it?

Reply:
not sure if this makes sense. something about the demand analysis + mfp composes weirdly for me and makes it hard to reason about whats happening

Reply:
We should probably eventually hit `mfp` with some `.optimize()` call once that gets written. Probably `remove_undemanded()` makes sense there! I have a few other optimizations ready to go once we land this (e.g. deduplicate expressions to reduce the size of `dummies`). Probably I should put a call in to `optimize()` and then make that route out to everything we know how to do with these.

Reply:
makes sense

Reply:
Actually, I think I'll punt on this for the moment, as this is not the last moment before any closure gets used. I.e. we'll peel off `before`s and end up with an `after`, each of which should be optimized then. We could optimize here too, but I don't think it is expected to do anything that the later optimizations wouldn't clean up.